### PR TITLE
Restructure the relationship between PipelineComponent and AlgorithmBase

### PR DESCRIPTION
**Pull request by ttung**
Restructure the relationship between PipelineComponent and AlgorithmBase

Rather than each specific PipelineComponent have a link to the AlgorithmBase that describes the contract that the algorithm must implement, each AlgorithmBase implementation points to the PipelineComponent that it serves.

The current arrangement forces all the algorithms to be defined and evaluated in the interpreter before the pipeline component is defined and evaluated.  If we want to create an algorithm that's only used in tests, this is not possible.

The new arrangement ensures that any implementation of a specific AlgorithmBase is registered with its corresponding PipelineComponent.

This adds explicit abstract classes to the AlgorithmBase class hierarchy to make it easier to detect what's an actual algorithm implementation.

Depends on #1094

Test plan: `make -j fast`

### CODE REVIEW

**Comment by ttung** [comment]
This also partially addresses #921

File: starfish/pipeline/algorithmbase.py [hunk 1]
```diff
@@ -1,17 +1,37 @@
...
 from starfish.imagestack.imagestack import ImageStack
 from starfish.intensity_table.intensity_table import IntensityTable
 from starfish.types import LOG
 from starfish.types._constants import STARFISH_EXTRAS_KEY
 from starfish.util.logging import LogEncoder
+from .pipelinecomponent import PipelineComponent
 
 
-class AlgorithmBaseType(type):
+DISALLOWED_BASES: Set[Type] = set()
```
Comment (shanaxel42) [inline starfish/pipeline/algorithmbase.py]: what's an example of a disallowed base?

Comment (shanaxel42) [inline starfish/pipeline/algorithmbase.py]: what's an example of a disallowed base?

Comment (ttung) [reply starfish/pipeline/algorithmbase.py]: Let me improve the documentation.  Basically, we want the subclasses of `FilterAlgorithmBase`, `RegistrationAlgorithmBase`, etc., and not the subclasses of `AlgorithmBase`.

Comment (ttung) [reply starfish/pipeline/algorithmbase.py]: I went with a different approach that removed the need for this.

**Pull request merged**